src/main/java/de/tum/in/www1/artemis/service/LearningGoalService.java

```diff
@@ -184,10 +184,39 @@ else if (!neighbor.isVisited() && vertexIsPartOfCycle(neighbor)) {
             var headVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getHeadLearningGoal().getTitle())).findFirst().orElseThrow();
             var tailVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getTailLearningGoal().getTitle())).findFirst().orElseThrow();
             // Only EXTENDS and ASSUMES are included in the generated graph as other relations are no problem if they are circular
+            // MATCHES relations are considered in the next step by merging the edges and combining the adjacencyLists
             switch (relation.getType()) {
                 case EXTENDS, ASSUMES -> graph.addEdge(tailVertex, headVertex);
             }
         }
+        // combine vertices that are connected through MATCHES
+        for (LearningGoalRelation relation : relations) {
+            if (relation.getType() == LearningGoalRelation.RelationType.MATCHES) {
+                var headVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getHeadLearningGoal().getTitle())).findFirst().orElseThrow();
+                var tailVertex = graph.vertices.stream().filter(vertex -> vertex.label.equals(relation.getTailLearningGoal().getTitle())).findFirst().orElseThrow();
+                if (headVertex.adjacencyList.contains(tailVertex) || tailVertex.adjacencyList.contains(headVertex)) {
+                    return true;
+                }
+                // create a merged vertex
+                var mergedVertex = new Vertex(tailVertex.label + ", " + headVertex.label);
+                // add all neighbours to merged vertex
+                for (Vertex vertex : headVertex.getAdjacencyList()) {
+                    mergedVertex.addNeighbor(vertex);
+                }
+                for (Vertex vertex : tailVertex.getAdjacencyList()) {
+                    mergedVertex.addNeighbor(vertex);
+                }
```

Comment:
Something like that:

mergedVertex.neighbours.addAll(headVertex.getAdjacencyList());
mergedVertex.neighbours.addAll(tailVertex.getAdjacencyList());